### DISCUSSION TITLE
how to install keyviz on linux (ubuntu)

### DISCUSSION THREAD

**Discussion by imppao** [mulaRahul/keyviz#306] [General]
**Describe the solution you'd like**
I would like to know how install keyviz on linux ubuntu, for simple linux user

**Additional context**
Thanks for your job

**Comment by Lucifer516-sudoer** [comment]
> **Describe the solution you'd like** I would like to know how install keyviz on linux ubuntu, for simple linux user
> 
> **Additional context** Thanks for your job

Just download the linux release from the releases, unzip it, then modify the file mode, then just run it, well as far as i know, it must be rendering a black screen for you.

or if you dont get what i said now, just run the following command in your terminal:
```shell
wget https://github.com/mulaRahul/keyviz/releases/download/v2.0.0a/keyviz_v2.0.0a_linux_x64.zip
unzip  keyviz_v2.0.0a_linux_x64.zip -d keyviz # unzips to the keyviz folder
cd keyviz
chmod +x keyviz # makes the keyviz file executable
./keyviz
```
Just in case if u get struck in blank window, press Alt+F4 to close that window ( i guess thats the normal way to close in most distros)

**Edit:**
I just forgot to say, if you are thinking of installing this in your system, you might have to store the inflated stuffs into ur bin folder and also add an desktop entry.
for further info, just google it. Google is your best friend.

**Reply by luifrancgom** [reply]
This is not a clear answer. #304 ask the same question. There is no  `README.txt` or `INSTALL.txt` to guide the user, at least in v2.0.0-alpha2

**Reply by luifrancgom** [reply]
Also making it executable I get a lot of errors even installing the requirements pointed out in the installation guide section of the repo.

**Reply by Lucifer516-sudoer** [reply]
> Also making it executable I get a lot of errors even installing the requirements pointed out in the installation guide section of the repo.

@luifrancgom can you please share the error messages that you are getting now. If I can, I am happy to help you. But if you don't share the issue with the community, there is no point in asking the question.

**Reply by luifrancgom** [reply]
You are right. Sorry if you think I was rude. My sincere apologies. Thank you for helping me. I will share the complete information:

## Information about the OS

Distributor ID:	Ubuntu
Description:	Ubuntu 24.04.2 LTS
Release:	24.04
Codename:	noble

## Keyviz information

Realease: [v2.0.0-alpha2](https://github.com/mulaRahul/keyviz/releases/tag/v2.0.0a2)

## Process trying to execute Keyviz

```bash
cd  ~/Downloads/
wget https://github.com/mulaRahul/keyviz/releases/download/v2.0.0a2/keyviz-v2.0.0a2-linux.zip
unzip keyviz-v2.0.0a2-linux.zip -d keyviz
cd keyviz
chmod +x keyviz
./keyviz
```
## Errors 

```bash
flutter: [ json:load ]: File: '/home/luifrancgom/Documents/config.json' not found!
flutter: registered mouse listener
flutter: keyboard listener registered
flutter: [ json:load ]: File: '/home/luifrancgom/Documents/style.json' not found!
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: MissingPluginException(No implementation found for method setIgnoreMouseEvents on channel window_manager)
#0      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:332)
<asynchronous suspension>
#1      WindowManager.setIgnoreMouseEvents (package:window_manager/src/window_manager.dart:699)
<asynchronous suspension>

[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: MissingPluginException(No implementation found for method setHasShadow on channel window_manager)
#0      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:332)
<asynchronous suspension>
#1      WindowManager.setHasShadow (package:window_manager/src/window_manager.dart:667)
<asynchronous suspension>
```

**Reply by Lucifer516-sudoer** [reply]
Thank you for sharing the detailed logs. I can see that you're encountering plugin-related errors when trying to launch Keyviz.

Based on the error messages, the application is missing implementations for the Flutter window_manager plugin, which is causing the crash. Additionally, there are configuration files that cannot be found.

Here are some troubleshooting steps you could try:

1. Install required dependencies:
   ```bash
   sudo apt update
   sudo apt install libgtk-3-dev libx11-dev pkg-config cmake ninja-build libblkid-dev libsecret-1-dev libjsoncpp-dev
   sudo apt-get install libayatana-appindicator3-dev # This was actually mentioned in the README
   ```

2. Create the missing configuration files:
   ```bash
   mkdir -p ~/Documents
   touch ~/Documents/config.json
   touch ~/Documents/style.json
   ```

3. Try running Keyviz with the --no-sandbox flag: (I am not sure if this will work)
   ```bash
   ./keyviz --no-sandbox
   ```
   

The errors suggest that Keyviz is looking for window_manager plugin implementations that aren't available on your system. This might be related to compatibility issues with Ubuntu 24.04, as the application was likely tested on earlier versions.

Please let me know if any of these steps help resolve the issue.

**Reply by luifrancgom** [reply]
I really appreciate your help but it didn't work in my case. Again thank you for your detail instructions. 

I already have the Documents folder but I understand the -p option

I created the files config.json and style.json

Also I installed the part about libgtk-3-dev libx11-dev pkg-config cmake ninja-build libblkid-dev libsecret-1-dev libjsoncpp-dev but I get the same message:

```bash
flutter: FormatException: Unexpected end of input (at character 1)
flutter: 
flutter: ^
flutter: 
flutter: [ json:load ]: File: '/home/luifrancgom/Documents/config.json' not found!
flutter: registered mouse listener
flutter: keyboard listener registered
flutter: FormatException: Unexpected end of input (at character 1)
flutter: 
flutter: ^
flutter: 
flutter: [ json:load ]: File: '/home/luifrancgom/Documents/style.json' not found!
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: MissingPluginException(No implementation found for method setIgnoreMouseEvents on channel window_manager)
#0      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:332)
<asynchronous suspension>
#1      WindowManager.setIgnoreMouseEvents (package:window_manager/src/window_manager.dart:699)
<asynchronous suspension>

[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: MissingPluginException(No implementation found for method setHasShadow on channel window_manager)
#0      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:332)
<asynchronous suspension>
#1      WindowManager.setHasShadow (package:window_manager/src/window_manager.dart:667)
<asynchronous suspension>
```
But again thank you for your help.

**Comment by Ten0** [comment]
Running it on Linux, I'm seeing the key presses on a plain black window that takes up the whole screen, so I can't see anything behind.

I suspect that is linked to me getting the following errors:
```
[ERROR:flutter/runtime/dart_vm_initializer.cc(40)] Unhandled Exception: MissingPluginException(No implementation found for method setIgnoreMouseEvents on channel window_manager)
#0      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:368)
<asynchronous suspension>
#1      WindowManager.setIgnoreMouseEvents (package:window_manager/src/window_manager.dart:699)
<asynchronous suspension>

[ERROR:flutter/runtime/dart_vm_initializer.cc(40)] Unhandled Exception: MissingPluginException(No implementation found for method setHasShadow on channel window_manager)
#0      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:368)
<asynchronous suspension>
#1      WindowManager.setHasShadow (package:window_manager/src/window_manager.dart:667)
<asynchronous suspension>
```

Do we actually know that anyone has successfully get this to work properly on linux, or is it plain broken?
https://github.com/mulaRahul/keyviz/issues/362

**Reply by Lucifer516-sudoer** [reply]
@Ten0 , I suppose that keyviz is not broken, rather it is not having the required implementations.

**Reply by Ten0** [reply]
If a program never works on a given platform, then yeah it's broken for that platform. 
I understand that this is currently mostly actionable by contributing the corresponding implementations on the underlying libraries.
But this project should stop pretending that Linux is supported if in its current state it's unusable. Or at least there should be a prominent notice on the installation journey, so that people don't lose time hopelessly trying to get it to work.

**Reply by Lucifer516-sudoer** [reply]
@Ten0  I agree with you. I hope keyviz will support linux in future, because right now I am being forced to switch to windows (which I hate to do).

**Comment by stephanschielke** [comment]
This is just an FYI on my test and experience of the Linux version

After reading https://github.com/mulaRahul/keyviz/pull/129#issuecomment-1515764213 and discovering the [alpha version 2.0.0](https://github.com/mulaRahul/keyviz/releases/tag/v2.0.0a3) with Linux support, I wanted to test it out on my Ubuntu Linux machine, but it was quite problematic getting the required dependencies. In the end, I built my own .deb file, which wasn't trivial either ([Fork](https://github.com/mulaRahul/keyviz/compare/main...stephanschielke:keyviz:main)).

Offering (or planning) Linux support is absolutely fantastic, and I don't want to demotivate you here; however, it feels like using Flutter for this is quite the uphill battle ahead, especially with some of the chosen plugins. I hope you'll find a path!

I discovered (and for now will use) [screenkey](https://gitlab.com/screenkey/screenkey) as a stable alternative for Linux users.
